lib/actions/usbsdfuTargetActions.js

```diff
@@ -268,103 +287,196 @@ function calculateSHA512Hash(image) {
-                dispatch(targetActions.initPacketSDKnownAction(firmwareImage.length));
-                initPacketParams = initPacketParams.set('sdSize', firmwareImage.length);
-                break;
-            case initPacket.FwType.SOFTDEVICE_BOOTLOADER:
-                logger.error('Firmware type BOOTLOADER_SOFTDEVICE is nor supported');
-                break;
-            default:
-                logger.error('Unknown firmware type');
-        }
-        const packet = await initPacket.createInitPacketUint8Array(initPacketParams);
+// Update firmware image and its length
+function handleImage(image, regions, memMap) {
+    if (!image.initPacket) {
+        throw new Error('Init packet was not created.');
+    }
 
-        const firmwareUpdates = new dfujs.DfuUpdates([{
-            initPacket: packet,
-            firmwareImage,
-        }]);
-        const serialTransport = new dfujs.DfuTransportSerial(port, 0);
-        const dfu = new dfujs.DfuOperation(firmwareUpdates, serialTransport);
+    const region = regions.find(r => r.name === image.name);
+    let fwSizeStr;
+
+    if (image.name.includes(RegionName.BOOTLOADER)) {
+        fwSizeStr = 'blSize';
+    } else if (image.name.includes(RegionName.SOFTDEVICE)) {
+        fwSizeStr = 'sdSize';
+    } else if (image.name.includes(RegionName.APPLICATION)) {
+        fwSizeStr = 'appSize';
+    } else {
+        throw new Error('Firmware type is unknown when setting firmware size.');
+    }
+
+    const firmwareImage = memMap.slicePad(region.startAddress, region.regionSize);
+    return image.set('firmwareImage', firmwareImage)
+                .setIn(['initPacket', fwSizeStr], firmwareImage.length);
+}
+
+// Update hardware version
+function handleHwVersion(image, hwVersion) {
+    return image.setIn(['initPacket', 'hwVersion'], hwVersion);
+}
+
+// Update SoftDevice required virsion
+function handleSdReq(image, fileMemMap, deviceInfo) {
+    const sdId = getSoftDeviceId(fileMemMap, deviceInfo);
+    let sdReq;
+
+    if (image.name.includes(RegionName.BOOTLOADER)) {
+        logger.info('SdReq for Bootloader is not detected. Need user input.');
+    } else if (image.name.includes(RegionName.SOFTDEVICE)) {
+        sdReq = [0x00];
+        logger.info('SdReq for SoftDevice is set as 0x00.');
+    } else if (image.name.includes(RegionName.APPLICATION)) {
+        sdReq = sdId ? [sdId] : undefined;
+        logger.info('SdReq for Application is not detected. Need user input.');
+    } else {
+        throw new Error('Firmware type is unknown when setting user input.');
+    }
 
-        dispatch(targetActions.targetWritableKnownAction(false));
-        logger.info('Performing DFU. This may take a few seconds');
+    return image.setIn(['initPacket', 'sdReq'], sdReq);
+}
+
+// Update init packet with ueser input
+function handleUserInput(imageInput, userInput) {
+    let image = imageInput;
+    let sdReq;
+    let fwVersion;
+
+    if (image.name.includes(RegionName.BOOTLOADER)) {
```

Comment:
What is the type/value of `image.name`? Is it not a string set from those `RegionName.*` values? I ask because I don't know if `.includes(..)` is intended or should it only be a simple string match (equality)?

Reply:
This is reserved for multiple Application. In case that there are multiple applications, we should name it with Application1 Application2 or something similar.

Reply:
Right, it could be stricter. So here it is using the image.name.includes(RegionName). RegionName.includes(image.name) cannot be used. We could find a better way later, since we dont have any multiple appliations test case for now.
Besides, I also suggest to chagne `MBR or Application` to `MBR` since if it is not a MBR then we wont see it any more.